blocks/basic/include/gnuradio-4.0/basic/FunctionGenerator.hpp

```diff
@@ -0,0 +1,297 @@
+                  Tag(850, createImpulseResponsePropertyMap(10.f, 20.f, .02f, .06f)) };
+@endcode
+
+)"">;
+
+template<typename T>
+    requires(std::floating_point<T>)
+struct FunctionGenerator : public gr::Block<FunctionGenerator<T>, BlockingIO<true>, FunctionGeneratorDoc> {
+    /**
+     * TODO: Taken from CtxSettings
+     * A predicate for matching two contexts
+     * The third "attempt" parameter indicates the current round of matching being done.
+     * This is useful for hierarchical matching schemes,
+     * e.g. in the first round the predicate could look for almost exact matches only,
+     * then in a a second round (attempt=1) it could be more forgiving, given that there are no exact matches available.
+     *
+     * The predicate will be called until it returns "true" (a match is found), or until it returns std::nullopt,
+     * which indicates that no matches were found and there is no chance of matching anything in a further round.
+     */
+    using MatchPredicate = std::function<std::optional<bool>(const property_map &, const property_map &, std::size_t)>;
+
+    PortIn<T>  in; // ClockSource input
+    PortOut<T> out;
+
+    Annotated<float, "sample_rate", Visible, Doc<"sample rate">>                        sample_rate = 1000.f;
+    Annotated<std::string, "signal_type", Visible, Doc<"see function_generator::Type">> signal_type = "Const";
+
+    // Parameters for different functions
+    Annotated<T, "start_value", Visible>                                               start_value    = T(0.);
+    Annotated<T, "final_value", Visible>                                               final_value    = T(0.);
+    Annotated<T, "duration", Visible, Doc<"in sec">>                                   duration       = T(0.);
+    Annotated<T, "round_off_time", Visible, Doc<"Specific to ParabolicRamp, in sec">>  round_off_time = T(0.);
+    Annotated<T, "impulse_time0", Visible, Doc<"Specific to ImpulseResponse, in sec">> impulse_time0  = T(0.);
+    Annotated<T, "impulse_time1", Visible, Doc<"Specific to ImpulseResponse, in sec">> impulse_time1  = T(0.);
+
+    Annotated<property_map, "trigger_meta_info"> trigger_meta_info{};
+    // TODO; need to use 2 std::vectors, can not use property_map as std::map key.
+    std::vector<property_map> function_keys{};   // matches the trigger tag map to the prestored settings
+    std::vector<property_map> function_values{}; // matches the trigger tag map to the prestored settings
+
+    MatchPredicate match_pred = [](auto, auto, auto) { return std::nullopt; };
+
+    T   _currentTime  = T(0.);
+    int sampleCounter = 0;
+
+private:
+    function_generator::Type _signalType = function_generator::parse(signal_type);
+    T                        _timeTick   = T(1.) / static_cast<T>(sample_rate);
+
+public:
+    void
+    settingsChanged(const property_map & /*old_settings*/, const property_map &new_settings) {
+        if (new_settings.contains("trigger_meta_info")) {
```

Comment:
A default-tag/enum-derived definition would help here instead of the literal string.